<a href="https://colab.research.google.com/github/map72ru/nlp/blob/main/hw_lesson_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/hw2/')

Mounted at /content/hw2/


In [2]:
import pandas as pd
import pickle as p
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk

In [3]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [4]:
with open('hw2/MyDrive/data/combine_df.pkl', 'rb') as f:
  combine_df = p.load(f)

sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
*   Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
*   Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
*	Исключим стоп-слова с помощью stop_words='english'.
*	Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().


In [5]:
vectorizer = CountVectorizer(max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=1000)
vectorizer2 = CountVectorizer(max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=1000)

In [6]:
combine_df['tweet_stemmed'].str.join(' ')

0            father dysfunct selfish drag kid dysfunct run
1        thank lyft credit use caus offer wheelchair va...
2                                           bihday majesti
3                  model love yoyou take yoyou time yoyour
4                                  factsguid societi motiv
                               ...                        
49154    thought factori left right polaris trump usele...
49155    feel like mermaid hairflip neverreadi formal w...
49156    hillari campaign today ohio omg amp use word l...
49157    happi work confer right mindset lead cultur de...
49158       song glad free download shoegaz newmus newsong
Name: tweet_stemmed, Length: 49159, dtype: object

In [7]:
vectors_stemmed = vectorizer.fit_transform(combine_df['tweet_stemmed'].str.join(' '))
vectors_lemmatized = vectorizer2.fit_transform(combine_df['tweet_lemmatized'].str.join(' '))

In [8]:
pd.DataFrame(vectors_stemmed.toarray(), columns = vectorizer.get_feature_names())

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


abl  absolut  accept  account  act  action  actor  actual  ad  adapt  \
0        0        0       0        0    0       0      0       0   0      0   
1        0        0       0        0    0       0      0       0   0      0   
2        0        0       0        0    0       0      0       0   0      0   
3        0        0       0        0    0       0      0       0   0      0   
4        0        0       0        0    0       0      0       0   0      0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ..    ...   
49154    0        0       0        0    0       0      0       0   0      0   
49155    0        0       0        0    0       0      0       0   0      0   
49156    0        0       0        0    0       0      0       0   0      0   
49157    0        0       0        0    0       0      0       0   0      0   
49158    0        0       0        0    0       0      0       0   0      0   

       ...  yo  yoga  york  young  youtub  yoyou  yoyour  yoyoy  yr  yummi  
0      ...   0     0     0      0       0      0       0      0   0      0  
1      ...   0     0     0      0       0      0       0      0   0      0  
2      ...   0     0     0      0       0      0       0      0   0      0  
3      ...   0     0     0      0       0      2       1      0   0      0  
4      ...   0     0     0      0       0      0       0      0   0      0  
...    ...  ..   ...   ...    ...     ...    ...     ...    ...  ..    ...  
49154  ...   0     0     0      0       0      0       0      0   0      0  
49155  ...   0     0     0      0       0      0       0      0   0      0  
49156  ...   0     0     0      0       0      0       0      0   0      0  
49157  ...   0     0     0      0       0      0       0      0   0      0  
49158  ...   0     0     0      0       0      0       0      0   0      0  

[49159 rows x 1000 columns]

In [9]:
pd.DataFrame(vectors_lemmatized.toarray(), columns = vectorizer2.get_feature_names())

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


able  absolutely  account  act  action  actor  actually  adapt  add  \
0         0           0        0    0       0      0         0      0    0   
1         0           0        0    0       0      0         0      0    0   
2         0           0        0    0       0      0         0      0    0   
3         0           0        0    0       0      0         0      0    0   
4         0           0        0    0       0      0         0      0    0   
...     ...         ...      ...  ...     ...    ...       ...    ...  ...   
49154     0           0        0    0       0      0         0      0    0   
49155     0           0        0    0       0      0         0      0    0   
49156     0           0        0    0       0      0         0      0    0   
49157     0           0        0    0       0      0         0      0    0   
49158     0           0        0    0       0      0         0      0    0   

       adventure  ...  yo  yoga  york  young  youtube  yoyou  yoyour  yoyoyou  \
0              0  ...   0     0     0      0        0      0       0        0   
1              0  ...   0     0     0      0        0      0       0        0   
2              0  ...   0     0     0      0        0      0       0        0   
3              0  ...   0     0     0      0        0      2       1        0   
4              0  ...   0     0     0      0        0      0       0        0   
...          ...  ...  ..   ...   ...    ...      ...    ...     ...      ...   
49154          0  ...   0     0     0      0        0      0       0        0   
49155          0  ...   0     0     0      0        0      0       0        0   
49156          0  ...   0     0     0      0        0      0       0        0   
49157          0  ...   0     0     0      0        0      0       0        0   
49158          0  ...   0     0     0      0        0      0       0        0   

       yr  yummy  
0       0      0  
1       0      0  
2       0      0  
3       0      0  
4       0      0  
...    ..    ...  
49154   0      0  
49155   0      0  
49156   0      0  
49157   0      0  
49158   0      0  

[49159 rows x 1000 columns]

Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
*	Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
*	Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
*	Исключим стоп-слова с помощью stop_words='english'.
*	Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().


In [10]:
tvectorizer = TfidfVectorizer(max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=1000)
tvectorizer2 = TfidfVectorizer(max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=1000)

In [11]:
tvectors_stemmed = tvectorizer.fit_transform(combine_df['tweet_stemmed'].str.join(' '))
tvectors_lemmatized = tvectorizer2.fit_transform(combine_df['tweet_lemmatized'].str.join(' '))

In [12]:
pd.DataFrame(tvectors_stemmed.toarray(), columns = tvectorizer.get_feature_names())

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


abl  absolut  accept  account  act  action  actor  actual   ad  adapt  \
0      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
1      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
2      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
3      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
4      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ...    ...   
49154  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49155  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49156  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49157  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49158  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   

       ...   yo  yoga  york  young  youtub     yoyou    yoyour  yoyoy   yr  \
0      ...  0.0   0.0   0.0    0.0     0.0  0.000000  0.000000    0.0  0.0   
1      ...  0.0   0.0   0.0    0.0     0.0  0.000000  0.000000    0.0  0.0   
2      ...  0.0   0.0   0.0    0.0     0.0  0.000000  0.000000    0.0  0.0   
3      ...  0.0   0.0   0.0    0.0     0.0  0.704816  0.371659    0.0  0.0   
4      ...  0.0   0.0   0.0    0.0     0.0  0.000000  0.000000    0.0  0.0   
...    ...  ...   ...   ...    ...     ...       ...       ...    ...  ...   
49154  ...  0.0   0.0   0.0    0.0     0.0  0.000000  0.000000    0.0  0.0   
49155  ...  0.0   0.0   0.0    0.0     0.0  0.000000  0.000000    0.0  0.0   
49156  ...  0.0   0.0   0.0    0.0     0.0  0.000000  0.000000    0.0  0.0   
49157  ...  0.0   0.0   0.0    0.0     0.0  0.000000  0.000000    0.0  0.0   
49158  ...  0.0   0.0   0.0    0.0     0.0  0.000000  0.000000    0.0  0.0   

       yummi  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0  
...      ...  
49154    0.0  
49155    0.0  
49156    0.0  
49157    0.0  
49158    0.0  

[49159 rows x 1000 columns]

In [13]:
pd.DataFrame(tvectors_lemmatized.toarray(), columns = tvectorizer2.get_feature_names())

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


able  absolutely  account  act  action  actor  actually  adapt  add  \
0       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
1       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
2       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
3       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
4       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
...     ...         ...      ...  ...     ...    ...       ...    ...  ...   
49154   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49155   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49156   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49157   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49158   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   

       adventure  ...   yo  yoga  york  young  youtube     yoyou    yoyour  \
0            0.0  ...  0.0   0.0   0.0    0.0      0.0  0.000000  0.000000   
1            0.0  ...  0.0   0.0   0.0    0.0      0.0  0.000000  0.000000   
2            0.0  ...  0.0   0.0   0.0    0.0      0.0  0.000000  0.000000   
3            0.0  ...  0.0   0.0   0.0    0.0      0.0  0.701493  0.370288   
4            0.0  ...  0.0   0.0   0.0    0.0      0.0  0.000000  0.000000   
...          ...  ...  ...   ...   ...    ...      ...       ...       ...   
49154        0.0  ...  0.0   0.0   0.0    0.0      0.0  0.000000  0.000000   
49155        0.0  ...  0.0   0.0   0.0    0.0      0.0  0.000000  0.000000   
49156        0.0  ...  0.0   0.0   0.0    0.0      0.0  0.000000  0.000000   
49157        0.0  ...  0.0   0.0   0.0    0.0      0.0  0.000000  0.000000   
49158        0.0  ...  0.0   0.0   0.0    0.0      0.0  0.000000  0.000000   

       yoyoyou   yr  yummy  
0          0.0  0.0    0.0  
1          0.0  0.0    0.0  
2          0.0  0.0    0.0  
3          0.0  0.0    0.0  
4          0.0  0.0    0.0  
...        ...  ...    ...  
49154      0.0  0.0    0.0  
49155      0.0  0.0    0.0  
49156      0.0  0.0    0.0  
49157      0.0  0.0    0.0  
49158      0.0  0.0    0.0  

[49159 rows x 1000 columns]

Проверьте ваши векторайзеры на корпусе который использовали на вебинаре, составьте
таблицу метод векторизации и скор который вы получили (в методах векторизации по
изменяйте параметры что бы добиться лучшего скора) обратите внимание как
падает/растёт скор при уменьшении количества фичей, и изменении параметров, так же
попробуйте применить к векторайзерам PCA для сокращения размерности посмотрите на
качество сделайте выводы

In [14]:
# Загружаем данные
data = open('hw2/MyDrive/data/corpus').read()
labels, texts = [], []
for i, line in enumerate(data.split("\n")):
    content = line.split()
    labels.append(content[0])
    texts.append(" ".join(content[1:]))

# создаем df
trainDF = pd.DataFrame()
trainDF['text'] = texts
trainDF['label'] = labels
trainDF.head(5)

text       label
0  Stuning even for the non-gamer: This sound tra...  __label__2
1  The best soundtrack ever to anything.: I'm rea...  __label__2
2  Amazing!: This soundtrack is my favorite music...  __label__2
3  Excellent Soundtrack: I truly like this soundt...  __label__2
4  Remember, Pull Your Jaw Off The Floor After He...  __label__2

In [16]:
from sklearn import model_selection, preprocessing, linear_model
from sklearn.metrics import accuracy_score

In [34]:
def calculate(vectorizer, trainDF):
  train_x, valid_x, train_y, valid_y = model_selection.train_test_split(trainDF['text'], trainDF['label'])

# labelEncode целевую переменную
  encoder = preprocessing.LabelEncoder()
  train_y = encoder.fit_transform(train_y)
  valid_y = encoder.fit_transform(valid_y)


  vectorizer.fit(trainDF['text'])
  xtrain_count =  vectorizer.transform(train_x)
  xvalid_count =  vectorizer.transform(valid_x)

  classifier = linear_model.LogisticRegression(solver='lbfgs', max_iter=1000)
  classifier.fit(xtrain_count, train_y)
  predictions = classifier.predict(xvalid_count)
  return accuracy_score(valid_y, predictions)


In [35]:
count_vect = CountVectorizer(analyzer='word', token_pattern=r'\w{1,}')
result = calculate(count_vect, trainDF)
print(result)


0.86


In [36]:
vectorizer = CountVectorizer(max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'))
result = calculate(vectorizer, trainDF)
print(result)


0.8436


In [37]:
vectorizer = CountVectorizer(max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=1000)
result = calculate(vectorizer, trainDF)
print(result)

0.8228


In [38]:
vectorizer = CountVectorizer(max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000)
result = calculate(vectorizer, trainDF)
print(result)

0.8504


In [39]:
vectorizer = CountVectorizer(max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000)
result = calculate(vectorizer, trainDF)
print(result)

0.8416


In [40]:
vectorizer = CountVectorizer(max_df=0.7, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000)
result = calculate(vectorizer, trainDF)
print(result)

0.8376


In [41]:
vectorizer = CountVectorizer(max_df=0.8, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000)
result = calculate(vectorizer, trainDF)
print(result)

0.8456


In [42]:
vectorizer = CountVectorizer(max_df=1, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000)
result = calculate(vectorizer, trainDF)
print(result)

0.4968


In [43]:
vectorizer = CountVectorizer(max_df=0.99, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000)
result = calculate(vectorizer, trainDF)
print(result)

0.8584


In [44]:
vectorizer = CountVectorizer(max_df=0.99999, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000)
result = calculate(vectorizer, trainDF)
print(result)

0.8456


In [77]:
vectorizer = CountVectorizer(max_df=0.99, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000)
result = calculate(vectorizer, trainDF)
print(result)

0.8608


In [45]:
apostrophe_dict = {
"ain't": "am not / are not",
"aren't": "are not / am not",
"can't": "cannot",
"can't've": "cannot have",
"'cause": "because",
"could've": "could have",
"couldn't": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"doesn't": "does not",
"don't": "do not",
"hadn't": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"haven't": "have not",
"he'd": "he had / he would",
"he'd've": "he would have",
"he'll": "he shall / he will",
"he'll've": "he shall have / he will have",
"he's": "he has / he is",
"how'd": "how did",
"how'd'y": "how do you",
"how'll": "how will",
"how's": "how has / how is",
"i'd": "I had / I would",
"i'd've": "I would have",
"i'll": "I shall / I will",
"i'll've": "I shall have / I will have",
"i'm": "I am",
"i've": "I have",
"isn't": "is not",
"it'd": "it had / it would",
"it'd've": "it would have",
"it'll": "it shall / it will",
"it'll've": "it shall have / it will have",
"it's": "it has / it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"mightn't've": "might not have",
"must've": "must have",
"mustn't": "must not",
"mustn't've": "must not have",
"needn't": "need not",
"needn't've": "need not have",
"o'clock": "of the clock",
"oughtn't": "ought not",
"oughtn't've": "ought not have",
"shan't": "shall not",
"sha'n't": "shall not",
"shan't've": "shall not have",
"she'd": "she had / she would",
"she'd've": "she would have",
"she'll": "she shall / she will",
"she'll've": "she shall have / she will have",
"she's": "she has / she is",
"should've": "should have",
"shouldn't": "should not",
"shouldn't've": "should not have",
"so've": "so have",
"so's": "so as / so is",
"that'd": "that would / that had",
"that'd've": "that would have",
"that's": "that has / that is",
"there'd": "there had / there would",
"there'd've": "there would have",
"there's": "there has / there is",
"they'd": "they had / they would",
"they'd've": "they would have",
"they'll": "they shall / they will",
"they'll've": "they shall have / they will have",
"they're": "they are",
"they've": "they have",
"to've": "to have",
"wasn't": "was not",
"we'd": "we had / we would",
"we'd've": "we would have",
"we'll": "we will",
"we'll've": "we will have",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what shall / what will",
"what'll've": "what shall have / what will have",
"what're": "what are",
"what's": "what has / what is",
"what've": "what have",
"when's": "when has / when is",
"when've": "when have",
"where'd": "where did",
"where's": "where has / where is",
"where've": "where have",
"who'll": "who shall / who will",
"who'll've": "who shall have / who will have",
"who's": "who has / who is",
"who've": "who have",
"why's": "why has / why is",
"why've": "why have",
"will've": "will have",
"won't": "will not",
"won't've": "will not have",
"would've": "would have",
"wouldn't": "would not",
"wouldn't've": "would not have",
"y'all": "you all",
"y'all'd": "you all would",
"y'all'd've": "you all would have",
"y'all're": "you all are",
"y'all've": "you all have",
"you'd": "you had / you would",
"you'd've": "you would have",
"you'll": "you shall / you will",
"you'll've": "you shall have / you will have",
"you're": "you are",
"you've": "you have"
}



short_word_dict = {
"121": "one to one",
"a/s/l": "age, sex, location",
"adn": "any day now",
"afaik": "as far as I know",
"afk": "away from keyboard",
"aight": "alright",
"alol": "actually laughing out loud",
"b4": "before",
"b4n": "bye for now",
"bak": "back at the keyboard",
"bf": "boyfriend",
"bff": "best friends forever",
"bfn": "bye for now",
"bg": "big grin",
"bta": "but then again",
"btw": "by the way",
"cid": "crying in disgrace",
"cnp": "continued in my next post",
"cp": "chat post",
"cu": "see you",
"cul": "see you later",
"cul8r": "see you later",
"cya": "bye",
"cyo": "see you online",
"dbau": "doing business as usual",
"fud": "fear, uncertainty, and doubt",
"fwiw": "for what it's worth",
"fyi": "for your information",
"g": "grin",
"g2g": "got to go",
"ga": "go ahead",
"gal": "get a life",
"gf": "girlfriend",
"gfn": "gone for now",
"gmbo": "giggling my butt off",
"gmta": "great minds think alike",
"h8": "hate",
"hagn": "have a good night",
"hdop": "help delete online predators",
"hhis": "hanging head in shame",
"iac": "in any case",
"ianal": "I am not a lawyer",
"ic": "I see",
"idk": "I don't know",
"imao": "in my arrogant opinion",
"imnsho": "in my not so humble opinion",
"imo": "in my opinion",
"iow": "in other words",
"ipn": "I’m posting naked",
"irl": "in real life",
"jk": "just kidding",
"l8r": "later",
"ld": "later, dude",
"ldr": "long distance relationship",
"llta": "lots and lots of thunderous applause",
"lmao": "laugh my ass off",
"lmirl": "let's meet in real life",
"lol": "laugh out loud",
"ltr": "longterm relationship",
"lulab": "love you like a brother",
"lulas": "love you like a sister",
"luv": "love",
"m/f": "male or female",
"m8": "mate",
"milf": "mother I would like to fuck",
"oll": "online love",
"omg": "oh my god",
"otoh": "on the other hand",
"pir": "parent in room",
"ppl": "people",
"r": "are",
"rofl": "roll on the floor laughing",
"rpg": "role playing games",
"ru": "are you",
"shid": "slaps head in disgust",
"somy": "sick of me yet",
"sot": "short of time",
"thanx": "thanks",
"thx": "thanks",
"ttyl": "talk to you later",
"u": "you",
"ur": "you are",
"uw": "you’re welcome",
"wb": "welcome back",
"wfm": "works for me",
"wibni": "wouldn't it be nice if",
"wtf": "what the fuck",
"wtg": "way to go",
"wtgp": "want to go private",
"ym": "young man",
"gr8": "great"
}


emoticon_dict = {
":)": "happy",
":‑)": "happy",
":-]": "happy",
":-3": "happy",
":->": "happy",
"8-)": "happy",
":-}": "happy",
":o)": "happy",
":c)": "happy",
":^)": "happy",
"=]": "happy",
"=)": "happy",
"<3": "happy",
":-(": "sad",
":(": "sad",
":c": "sad",
":<": "sad",
":[": "sad",
">:[": "sad",
":{": "sad",
">:(": "sad",
":-c": "sad",
":-< ": "sad",
":-[": "sad",
":-||": "sad"
}

In [46]:
def replacer(value: str, vocabulary: dict) -> str:
  word = value
  for token in value.split():
    if token in vocabulary.keys():
       word = word.replace(token, vocabulary[token])
  return word

In [47]:
import re
data = trainDF.copy()
data['text'] = data.apply(lambda x: x['text'].lower(), axis=1)
data['text'] = data.apply(lambda x: replacer(x['text'], apostrophe_dict), axis=1)
data['text'] = data.apply(lambda x: replacer(x['text'], short_word_dict), axis=1)
data['text'] = data.apply(lambda x: replacer(x['text'], emoticon_dict), axis=1)
data['text'] = data.apply(lambda x: re.sub(r'[^\w\s]', ' ', x['text']), axis=1)
data['text'] = data.apply(lambda x: re.sub(r'[^a-zA-Z0-9]', ' ', x['text']), axis=1)
data['text'] = data.apply(lambda x: re.sub(r'[^a-zA-Z]', ' ', x['text']), axis=1)
data['text'] = data.apply(lambda x: ' '.join([w for w in x['text'].split() if len(w)>1]), axis=1)

In [72]:
nltk.download('wordnet')
data1 = data.copy()
ps = nltk.stem.wordnet.WordNetLemmatizer()
data1['text'] = data1.apply(lambda x: [ps.lemmatize(w) for w in x['text'].split()], axis=1)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [74]:
data1['text'] = data1.apply(lambda x: ' '.join(x['text']), axis=1)

In [75]:
vectorizer = CountVectorizer(max_df=0.99, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000)
result = calculate(vectorizer, data)
print(result)

0.8472


In [76]:
vectorizer = CountVectorizer(max_df=0.99, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000)
result = calculate(vectorizer, data1)
print(result)

0.8504


Score - Параметры
*   0.86 - analyzer='word', token_pattern=r'\w{1,}'
*   0.8436 - max_df=0.9, stop_words=nltk.corpus.stopwords.words('english')
*   0.8228 - max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=1000
*   0.8504 - max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000
*   0.8416 - max_df=0.9, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000
*   0.8376 - max_df=0.7, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000
*   0.8456 - ax_df=0.8, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000
*   0.4968 - max_df=1, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000
*   0.8584 - max_df=0.99, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000
*   0.8556 - max_df=0.99999, stop_words=nltk.corpus.stopwords.words('english'), max_features=30000
*   0.8606 - max_df=0.99, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000
*   0.8472 - max_df=0.99, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000 + предобработка текста
*   0,8504 - max_df=0.99, stop_words=nltk.corpus.stopwords.words('english'), max_features=10000 + предобработка текста + лемитизация

Вывод: как ни странно, но лучше ничего особо не делать и получается чуть ли не самый лучший результат

<table>
</table>